core/client/fs/src/main/java/alluxio/master/MasterHealthCheckClient.java

```diff
@@ -0,0 +1,184 @@
+     */
+    public HealthCheckClient build() {
+      return new MasterHealthCheckClient(mAlluxioMasterName, mProcessCheck);
+    }
+  }
+
+  /**
+   * Runnable for checking if the AlluxioMaster is serving RPCs.
+   */
+  public final class MasterServingCheckRunnable implements Runnable {
+
+    @Override
+    public void run() {
+      try {
+        LOG.debug("Checking master is serving requests");
+        FileSystem fs = FileSystem.Factory.get();
+        fs.exists(new AlluxioURI("/"));
+        LOG.debug("Master is serving requests");
+      } catch (Throwable e) {
+        throw new RuntimeException(e);
+      }
+    }
+  }
+
+  /**
+   * Runnable for checking if the AlluxioMaster process are running in all the masters hosts.
+   * This include both primary and stand-by masters.
+   */
+  public final class ProcessCheckRunnable implements Runnable {
+    private String mAlluxioMasterName;
+
+    /**
+     * Creates a new instance of ProcessCheckRunnable.
+     *
+     * @param alluxioMasterName the Alluxio master process name
+     */
+    public ProcessCheckRunnable(String alluxioMasterName) {
+      mAlluxioMasterName = alluxioMasterName;
+    }
+
+    @Override
+    public void run() {
+      MasterInquireClient client = MasterInquireClient.Factory.create();
+      try {
+        while (true) {
+          List<InetSocketAddress> addresses = client.getMasterRpcAddresses();
+          for (InetSocketAddress address : addresses) {
+            String host = address.getHostName();
+            LOG.debug("Master health check on node {}", host);
+            String cmd = String.format("ssh %s %s %s", ShellUtils.COMMON_SSH_OPTS, host,
+                    "ps -ef | grep \"" + mAlluxioMasterName + "$\" | "
+                    + "grep \"java\" | "
+                    + "awk '{ print $2; }'");
+            LOG.debug("Executing: {}", cmd);
+            String output = ShellUtils.execCommand("bash", "-c", cmd);
+            if (output.isEmpty()) {
+              throw new IllegalStateException(
+                      String.format("Master process is not running on the host %s", host));
+            }
+            LOG.debug("Master running on node {} with pid={}", host, output);
+          }
+          CommonUtils.sleepMs(Constants.SECOND_MS);
+        }
+      } catch (Throwable e) {
+        LOG.error("Exception thrown in the master process check {}", e);
+        throw new RuntimeException(e);
+      }
+    }
+  }
+
+  /**
+   * Creates a new instance of MasterHealthCheckClient.
+   *
+   * @param alluxioMasterName the Alluxio master process name
+   * @param processCheck whether to check the AlluxioMaster process is alive
+   */
+  public MasterHealthCheckClient(String alluxioMasterName, boolean processCheck) {
+    mAlluxioMasterName = alluxioMasterName;
+    mProcessCheck = processCheck;
+  }
+
+  @Override
+  public boolean isServing() {
+    ExecutorService mExecutorService = Executors.newFixedThreadPool(2);
```

Comment:
is it intended here to create the thread pool on method, not in the constructor?
I am curious b/c say in the future, we have multiple methods for health check, do we maintain thread pools per method?

Reply:
At the moment may not be a problem since we are calling the ```MasterHealthCheckClient``` exclusively from the ```AlluxioMasterMonitor``` which it's only called once and terminate. But as you mentioned, if that usage pattern changes that may become a possible issue, I am going to move the thread pool to the constructor, thanks!